lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -200,6 +205,19 @@ static OptionalParseResult customTypeParser(AsmParser &parser, StringRef name,
     return result = FVectorType::get(elementType, width), success();
   }
 
+  if (name.equals("ref")) {
+    FIRRTLBaseType type;
+    if (parser.parseLess() || parseNestedBaseType(type, parser) ||
+        parser.parseGreater())
+      return failure();
+
+    if (!type.isGround())
+      return parser.emitError(parser.getNameLoc(),
+                              "reference type must be ground");
```

Comment:
This seems like something that a verifier should be checking and not the parser. I also think that just removing this will have the same effect.

Reply:
> I also think that just removing this will have the same effect.

Not sure what you mean, sorry.  Removing it changes behavior in this way: the assertion in `RefType::get` will fail, and if I remove the assertion *and* this check, no error is produced ("must be ground" test in `errors.mlir`).

As I understand it, I think this is because we expose our C++ types (Type's)  in TableGen as type constraints--- so operations using a constraint type have the corresponding predicate verified (`CPred`s, etc.), but if they just take a `FIRRTLType` (or `FIRRTLBaseType` or whatever) then any derived type's checks (attached to the "corresponding" constraint type) won't be checked.  For example, any predicates included on the tablegen "RefType" aren't checked for module ports that take `FIRRTLType`, even when they're `RefType`, only the constraints on `FIRRTLType`'s tablegen type are checked.

Here, the `RefType` in TableGen could check this (instead of just `isa<RefType>`) but it'd only apply to TableGen-defined things that make use of this type explicitly (like the future ops, but not a module port for example).

The only other type verification we do in FIRRTL during parsing, seems to be regarding negative width integers, so in FIRRTL not a lot of previous work.  HWTypes has parsers that check their verifiers during construction (`ArrayType`) using the C++ `::verify` hook.

> This seems like something that a verifier should be checking and not the parser

We don't have verifiers for this or other C++ types in FIRRTL (`Type::verify`), but that's a good idea and probably where this logic should live :+1:.  Will update with that shortly.

Unfortunately just adding the type verifier isn't enough: StorageUniquer has an [assertion failure if you attempt to construct a Type that fails verification](https://github.com/llvm/llvm-project/blob/42f32f69ade78fe443c911cdbacc1d77fbb3d843/mlir/include/mlir/IR/StorageUniquerSupport.h#L153).  So still need to check for valid types during construction I think, but at least the logic will live in a single place.  In the future I think we will want to loosen this from ground-only to any passive type.